src/cdomains/octApronDomain.ml

```diff
@@ -0,0 +1,611 @@
+open Prelude
+open Cil
+open Pretty
+(* For Apron implementation of octagons *)
+open Apron
+
+module Man =
+struct
+  (* Manager type, parameter for the command below *)
+  type mt = Oct.t
+  (* A type of manager allocated by the underlying octagon domain *)
+  type t = mt Manager.t
+
+  (* Allocate a new manager to manipulate octagons *)
+  let mgr = Oct.manager_alloc ()
+  (* Making an environment from a set of integer and real variables. 
+  Raise Failure in case of name conflict. 
+  In this case the environment is empty to begin with. *)
+  let eenv = Environment.make [||] [||]
+end
+
+(* Generic operations on abstract values at level 1 of interface, there is also Abstract0 *)
+module A = Abstract1 
+
+module D =
+struct
+  type t = Man.mt A.t
+
+  let name () = "APRON octagon numerical abstract domain"
+
+  let topE = A.top    Man.mgr
+  let botE = A.bottom Man.mgr
+
+  let top () = topE Man.eenv
+  let bot () = botE Man.eenv
+  let is_top = A.is_top    Man.mgr
+  let is_bot = A.is_bottom Man.mgr
+
+  let to_yojson x = failwith "TODO implement to_yojson"
+  let invariant _ _ = Invariant.none
+  let tag _ = failwith "Std: no tag"
+  let arbitrary () = failwith "no arbitrary"
+
+  let short n (x:t) =
+    A.print Legacy.Format.str_formatter x;
+    Legacy.Format.flush_str_formatter ()
+
+  let print_lincons l = Lincons0.print string_of_int Format.std_formatter l
+  let print_expression x = print_endline (Pretty.sprint 20 (Cil.d_exp () x))
+  let print_octagon o = print_endline (short 30 o)
+
+  let join x y =
+    let ret = if is_bot x then
+      y
+    else if is_bot y then
+      x
+    (* This would make sense, but causes:
+      Fatal error: exception Apron.Manager.Error(_) *)
+    (* else if is_top x || is_top y then
+      top () *) 
+    else
+      A.join (Man.mgr) x y in
```

Comment:
Why is it necessary to special case the `is_bot` and `is_top` things? Can `A.join` not handle them even though `A.is_bottom` and `A.is_top` also come from Apron?

Reply:
Same for `meet` and `widen`. Or is this a performance optimization?

Reply:
This logic originates from poly. I thought it might be some performance optimization, but if I just return `A.join (Man.mgr) x y` then I get `Fatal error: exception Apron.Manager.Error(_)`, which is not explaining the exception very well. The problem is when the apron manager tries to join something with bottom. If it joins top and bottom, only then it works.

Reply:
According to Apron documentation there should be some information with that exception: https://antoinemine.github.io/Apron/doc/api/ocaml/Manager.html#EXCEPTIONError. I suppose you could try catching it and using `print_exclog` to display it. Maybe that explains something.

Reply:
Thank you for the suggestion. Printing the exclog explained the reason for the Error. the error message was: `"The two abstract values have different environments."` That exception comes from [here](https://github.com/antoinemine/apron/blob/28e9c36fe6f64c2365b2535ca28f55673d7858e9/apron/ap_abstract1.c#L106). The environments are compared [here](https://github.com/antoinemine/apron/blob/28e9c36fe6f64c2365b2535ca28f55673d7858e9/apron/ap_environment.c#L657). It happens in `32-widen-context/01-on.c` in line 12 where an octagon with constraints (its environment is filled with variables) is joined with an octagon which is `bottom` (has an empty environment). I've talked about it with Michael and the conclusion was that this happens when we join with dead code.
This same if-based structure with `is_top` and `is_bottom` is there for `join`, `meet`, `widen`, `narrow`, `equal`, `leq` for the same reason: to avoid doing an operation with octagons who have different environments. This overhead code could be:

1. removed and only the Apron function would be used, but then some sneaky hacks need to be made to have equal environments (maybe adding the context from myCFG.current_node to bottom before joining), so the code would be shorter, but maybe less readable/logical
2. moved to so util function to shorten the code at least

Any advice about what to do @sim642 ?

Reply:
Since it's necessary, I think it's fine to keep it this way, but there should at least be a comment above explaining why (`bot` has an octagon with incompatible environment or whatever).

Also a good idea might be to register the special Apron exception printer with OCaml's exception printing: https://ocaml.org/learn/tutorials/error_handling.html#Printing. That way it should be possible to make unhandled Apron exceptions automatically print our all their information without having to put any try-with constructs all over these operations.

Reply:
The comment has been added and the suggestion about registering a printer is moved to #237.